### PR TITLE
+ht2 validate that WINDOW_UPDATE must be > 0

### PR DESCRIPTION
**Pull request by ktoso**
Validating window value, small little addition that I added while I was looking at flow control handling.

### CODE REVIEW

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 1]
```diff
@@ -472,6 +472,22 @@ class Http2ServerSpec extends AkkaSpec with WithInPendingUntilFixed with Eventua
...
+        errorCode should ===(ErrorCode.PROTOCOL_ERROR)
+      }
+      "reject WINDOW_UPDATE for stream with zero increment with PROTOCOL_ERROR" in new TestSetup with RequestResponseProbes {
+        // making sure we don't handle stream 0 and others differently here
+        sendHEADERS(1, endStream = false, endHeaders = true, HPackSpecExamples.C41FirstRequestWithHuffman)
+        sendWINDOW_UPDATE(1, 0) // illegal
+        val (_, errorCode) = expectGOAWAY()
+
+        errorCode should ===(ErrorCode.PROTOCOL_ERROR)
+      }
```
Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: for sanity / future resilience checking both cases, even though we handle it in one spot.

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/framing/Http2FrameParsing.scala [hunk 2]
```diff
@@ -117,8 +117,9 @@ private[http2] class Http2FrameParsing(shouldReadPreface: Boolean) extends ByteS
             // TODO: check frame size
             // TODO: check flags
             // TODO: check reserved flag
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/framing/Http2FrameParsing.scala]: You can remove this as well. The "reserved flag" is actually the sign when we read a complete `readIntBE`. So, technically the error message should then also be "Reserved flag in WINDOW_UPDATE" must be 0.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 3]
```diff
@@ -472,6 +472,22 @@ class Http2ServerSpec extends AkkaSpec with WithInPendingUntilFixed with Eventua
...
+        sendWINDOW_UPDATE(0, 0) // illegal
+        val (_, errorCode) = expectGOAWAY()
+
+        errorCode should ===(ErrorCode.PROTOCOL_ERROR)
+      }
+      "reject WINDOW_UPDATE for stream with zero increment with PROTOCOL_ERROR" in new TestSetup with RequestResponseProbes {
+        // making sure we don't handle stream 0 and others differently here
+        sendHEADERS(1, endStream = false, endHeaders = true, HPackSpecExamples.C41FirstRequestWithHuffman)
+        sendWINDOW_UPDATE(1, 0) // illegal
+        val (_, errorCode) = expectGOAWAY()
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: Should be RST_STREAM here. No need to kill the connection when only one substream is affected.

6.9:

```
   A receiver MUST treat the receipt of a WINDOW_UPDATE frame with an
   flow-control window increment of 0 as a stream error (Section 5.4.2)
   of type PROTOCOL_ERROR; errors on the connection flow-control window
   MUST be treated as a connection error (Section 5.4.1).
```

Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: Oh, good catch thanks.

Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: Done, here we get a RST_STREAM now, handled rather nicely IMO with new super type for exceptions on stream level, which carry streamid

**Comment by ktoso** [comment]
Should be ready, please review

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Compliance.scala [hunk 4]
```diff
@@ -38,10 +38,16 @@ private[akka] object Http2Compliance {
   }
 
   class Http2ProtocolException(val errorCode: ErrorCode, message: String) extends IllegalStateException(message)
+  class Http2ProtocolStreamException(val streamId: Int, val errorCode: ErrorCode, message: String) extends IllegalStateException(message)
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Compliance.scala]: LGTM

**Pull request merged**